clientConn.go

```diff
@@ -77,13 +84,17 @@ func (c *ClientConn) readResponse() segment {
 		return outMsg
 	case <-time.After(c.ReadTimeout):
 		if c.lastError == nil {
+			c.lock.Lock()
 			c.lastError = fmt.Errorf("Timeout waiting for message response")
+			c.lock.Unlock()
 		}
 		return segment{}
 	}
 }
 
 func (c *ClientConn) RegisterEvent(name string, handler func(response map[string]interface{})) (err error) {
+	c.lock.Lock()
```

Comment:
Dead lock RegisterEvent and readResponse.(lock twice to the some sync.RWMutex)
You need two locks here.

Reply:
Going into fixing this - i really see the need to refactor the usage of last error as channel. It might considered complexity but mutex are a much bigger pain to solve in this case.